### PR TITLE
Add support for local machine ID

### PR DESCRIPTION
**Pull request by daniel-de-vera**
This is in the context of issue https://github.com/signadot/signadot/issues/3837.
Runs on top of the following PRs: https://github.com/signadot/signadot/pull/3857, https://github.com/signadot/libconnect/pull/38, https://github.com/signadot/go-sdk/pull/36

### CODE REVIEW

**Review by scott-cotton — COMMENTED** [review]
started the review.

Some comments/questions:

- Should we remove the api related stuff from internal/config/locald.go?
- Should we remove the grpc/go-sdk model conversions?
- Should we detect operator version and error out on old operator appropriately (since local is beta)?

**Comment by daniel-de-vera** [comment]
> started the review.
> 
> Some comments/questions:
> 
> * Should we remove the api related stuff from internal/config/locald.go?

Not sure, I understand what you mean by that, can you expand?

> * Should we remove the grpc/go-sdk model conversions?

Do you mean across the whole CLI? (what would look like a massive change to me).

> * Should we detect operator version and error out on old operator appropriately (since local is beta)?

I think that's a good idea.

**Comment by scott-cotton** [comment]
> > * Should we remove the grpc/go-sdk model conversions?
> 
> Do you mean across the whole CLI? (what would look like a massive change to me).
> 

I renamed these funcs:
```
23-11-09 scott@air cli % git diff | grep func
-func ToModelsSandboxSpec(grpcSpec *structpb.Struct) (*models.SandboxSpec, error) {
+func xToModelsSandboxSpec(grpcSpec *structpb.Struct) (*models.SandboxSpec, error) {
@@ -23,7 +23,7 @@ func ToModelsSandboxSpec(grpcSpec *structpb.Struct) (*models.SandboxSpec, error)
-func ToGRPCSandbox(sb *models.Sandbox) (*structpb.Struct, error) {
+func xToGRPCSandbox(sb *models.Sandbox) (*structpb.Struct, error) {
@@ -32,7 +32,7 @@ func ToGRPCSandbox(sb *models.Sandbox) (*structpb.Struct, error) {
-func ToGRPCSandboxSpec(sbs *models.SandboxSpec) (*structpb.Struct, error) {
+func xToGRPCSandboxSpec(sbs *models.SandboxSpec) (*structpb.Struct, error) {
@@ -42,7 +42,7 @@ func ToGRPCSandboxSpec(sbs *models.SandboxSpec) (*structpb.Struct, error) {
-func ToModelsSandbox(grpcSandbox *structpb.Struct) (*models.Sandbox, error) {
+func xToModelsSandbox(grpcSandbox *structpb.Struct) (*models.Sandbox, error) {
```
in `internal/locald/api/sandboxmanager/convert` and it compiles fine, so they are now unused.

**Comment by scott-cotton** [comment]
> > * Should we remove the api related stuff from internal/config/locald.go?
> 
> Not sure, I understand what you mean by that, can you expand?
> 

Sorry, it was too vague.  Was thinking that with the removal of sandbox apply from sandbox manager, we would probably not need the api key nor initialise APIConfig inside of anything in `locald`.  I thought I remembered adding this CIConfig api support specifically for the sandbox apply proxy, and greping around it looks to me like we can remove APIKey from the CIConfig struct, as well as related code setting up the config for a `*manager` to communicate with the API.

wdyt?

**Comment by daniel-de-vera** [comment]
@scott-cotton, I implemented this:

> Should we detect operator version and error out on old operator appropriately (since local is beta)?

In PR https://github.com/signadot/cli/pull/92

**Comment by daniel-de-vera** [comment]
> > > * Should we remove the grpc/go-sdk model conversions?
> > 
> > 
> > Do you mean across the whole CLI? (what would look like a massive change to me).
> 
> I renamed these funcs:
> 
> ```
> 23-11-09 scott@air cli % git diff | grep func
> -func ToModelsSandboxSpec(grpcSpec *structpb.Struct) (*models.SandboxSpec, error) {
> +func xToModelsSandboxSpec(grpcSpec *structpb.Struct) (*models.SandboxSpec, error) {
> @@ -23,7 +23,7 @@ func ToModelsSandboxSpec(grpcSpec *structpb.Struct) (*models.SandboxSpec, error)
> -func ToGRPCSandbox(sb *models.Sandbox) (*structpb.Struct, error) {
> +func xToGRPCSandbox(sb *models.Sandbox) (*structpb.Struct, error) {
> @@ -32,7 +32,7 @@ func ToGRPCSandbox(sb *models.Sandbox) (*structpb.Struct, error) {
> -func ToGRPCSandboxSpec(sbs *models.SandboxSpec) (*structpb.Struct, error) {
> +func xToGRPCSandboxSpec(sbs *models.SandboxSpec) (*structpb.Struct, error) {
> @@ -42,7 +42,7 @@ func ToGRPCSandboxSpec(sbs *models.SandboxSpec) (*structpb.Struct, error) {
> -func ToModelsSandbox(grpcSandbox *structpb.Struct) (*models.Sandbox, error) {
> +func xToModelsSandbox(grpcSandbox *structpb.Struct) (*models.Sandbox, error) {
> ```
> 
> in `internal/locald/api/sandboxmanager/convert` and it compiles fine, so they are now unused.

Oh, I see! It's done now.

**Comment by daniel-de-vera** [comment]
> > > * Should we remove the api related stuff from internal/config/locald.go?
> > 
> > 
> > Not sure, I understand what you mean by that, can you expand?
> 
> Sorry, it was too vague. Was thinking that with the removal of sandbox apply from sandbox manager, we would probably not need the api key nor initialise APIConfig inside of anything in `locald`. I thought I remembered adding this CIConfig api support specifically for the sandbox apply proxy, and greping around it looks to me like we can remove APIKey from the CIConfig struct, as well as related code setting up the config for a `*manager` to communicate with the API.
> 
> wdyt?

Got it, yes, looks like a good idea.
I will check it and remove unneeded things.

**Comment by daniel-de-vera** [comment]
> > > * Should we remove the api related stuff from internal/config/locald.go?
> > 
> > 
> > Not sure, I understand what you mean by that, can you expand?
> 
> Sorry, it was too vague. Was thinking that with the removal of sandbox apply from sandbox manager, we would probably not need the api key nor initialise APIConfig inside of anything in `locald`. I thought I remembered adding this CIConfig api support specifically for the sandbox apply proxy, and greping around it looks to me like we can remove APIKey from the CIConfig struct, as well as related code setting up the config for a `*manager` to communicate with the API.
> 
> wdyt?

@scott-cotton, this is done now, in https://github.com/signadot/cli/pull/91/commits/eaf51d88248e472a8cafec20433975ce1a1cd6cc

File: internal/locald/sandboxmanager/sandbox_controller.go [hunk 1]
```diff
@@ -43,6 +43,8 @@ func newSBController(log *slog.Logger, sandbox *tunapiv1.WatchLocalSandboxesResp
 	}
 	// run the controller
 	go ctrl.run()
+	// trigger a reconcile
+	ctrl.triggerReconcile()
```
Comment (scott-cotton) [inline internal/locald/sandboxmanager/sandbox_controller.go:47]: I would put this in `ctrl.run()` so it is more self-contained.

File: internal/locald/sandboxmanager/sandbox_controller.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+func (ctrl *sbController) closeRevTunnel(xwName string) {
+	revtun := ctrl.revtuns[xwName]
+	if revtun == nil {
+		return
+	}
+
+	select {
+	case <-revtun.rtToClose:
+	default:
+		ctrl.log.Debug("sandbox controller closing revtun", "local", xwName)
```
Comment (scott-cotton) [inline internal/locald/sandboxmanager/sandbox_controller.go:207]: I think that, in order for the select to protect against double closures, we would need to put the call to `ctrl.log.Debug` after the `close()`

File: internal/locald/sandboxmanager/sandboxes_watcher.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+		wg.Add(1)
+		// overwrite the delete function
+		sbMon.delFn = func() {
+			defer wg.Done()
+		}
+		// stop the sandbox controller
+		sbMon.stop()
+	}
+	sbw.sbMu.Unlock()
+
```
Comment (scott-cotton) [inline internal/locald/sandboxmanager/sandboxes_watcher.go:170]: This code was hard to understand, maybe some comments would help:
- the grpcserver has called graceful stop, so no new sandboxes will be processed during a call to this function
- the delFn is called at the end of the sb monitor run, so the wg will wait for all monitors to stop running.
-

**Review by scott-cotton — APPROVED** [review]
Approving, left some minor suggestions in second review round.

File: internal/locald/sandboxmanager/sandboxes_watcher.go (hunk 3 above)
Comment (daniel-de-vera) [reply internal/locald/sandboxmanager/sandboxes_watcher.go:178]: I addressed this in https://github.com/signadot/cli/pull/94

File: internal/locald/sandboxmanager/sandbox_controller.go (hunk 2 above)
Comment (daniel-de-vera) [reply internal/locald/sandboxmanager/sandbox_controller.go:207]: I addressed this in https://github.com/signadot/cli/pull/94

File: internal/locald/sandboxmanager/sandbox_controller.go (hunk 1 above)
Comment (daniel-de-vera) [reply internal/locald/sandboxmanager/sandbox_controller.go:47]: I addressed this in https://github.com/signadot/cli/pull/94

**Pull request merged**